GroupMappersTest test fails in keycloak-client

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

The test in keycloak-client the test `GroupMappersTest` fails in a few tests after the changes in #31842. For example `testGroupMappers`:

```
2024-11-08T10:39:56.1298643Z [ERROR] org.keycloak.client.testsuite.group.GroupMappersTest.testGroupMappers  Time elapsed: 2.497 s  <<< ERROR!
2024-11-08T10:39:56.1299997Z java.lang.RuntimeException: 
2024-11-08T10:39:56.1301219Z com.fasterxml.jackson.databind.JsonMappingException: Illegal state. Please init first before obtaining provider
2024-11-08T10:39:56.1309611Z  at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1005] (through reference chain: org.keycloak.jose.jwk.JSONWebKeySet["keys"]->java.lang.Object[][0]->org.keycloak.jose.jwk.JWK["x5c"])
2024-11-08T10:39:56.1311208Z 	at org.keycloak.client.testsuite.common.OAuthClient.getRealmKeys(OAuthClient.java:644)
2024-11-08T10:39:56.1312097Z 	at org.keycloak.client.testsuite.common.OAuthClient.getRealmPublicKey(OAuthClient.java:615)
2024-11-08T10:39:56.1312967Z 	at org.keycloak.client.testsuite.common.OAuthClient.verifyToken(OAuthClient.java:595)
2024-11-08T10:39:56.1313806Z 	at org.keycloak.client.testsuite.group.AbstractGroupTest.login(AbstractGroupTest.java:61)
2024-11-08T10:39:56.1314744Z 	at org.keycloak.client.testsuite.group.GroupMappersTest.testGroupMappers(GroupMappersTest.java:138)
2024-11-08T10:39:56.1315533Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2024-11-08T10:39:56.1316072Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2024-11-08T10:39:56.1316597Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2024-11-08T10:39:56.1317432Z Caused by: com.fasterxml.jackson.databind.JsonMappingException: Illegal state. Please init first before obtaining provider
2024-11-08T10:39:56.1319083Z  at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1005] (through reference chain: org.keycloak.jose.jwk.JSONWebKeySet["keys"]->java.lang.Object[][0]->org.keycloak.jose.jwk.JWK["x5c"])
2024-11-08T10:39:56.1320457Z 	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:276)
2024-11-08T10:39:56.1321405Z 	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:630)
2024-11-08T10:39:56.1322640Z 	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:618)
2024-11-08T10:39:56.1323648Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:143)
2024-11-08T10:39:56.1324650Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:310)
2024-11-08T10:39:56.1325611Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
2024-11-08T10:39:56.1326634Z 	at com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:218)
2024-11-08T10:39:56.1327698Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
2024-11-08T10:39:56.1328819Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:310)
2024-11-08T10:39:56.1329767Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
2024-11-08T10:39:56.1331148Z 	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342)
2024-11-08T10:39:56.1332151Z 	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4905)
2024-11-08T10:39:56.1332931Z 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3885)
2024-11-08T10:39:56.1333657Z 	at org.keycloak.util.JsonSerialization.readValue(JsonSerialization.java:98)
2024-11-08T10:39:56.1334355Z 	at org.keycloak.util.JsonSerialization.readValue(JsonSerialization.java:83)
2024-11-08T10:39:56.1335230Z 	at org.keycloak.client.testsuite.common.OAuthClient.getRealmKeys(OAuthClient.java:642)
2024-11-08T10:39:56.1335794Z 	... 7 more
2024-11-08T10:39:56.1336320Z Caused by: java.lang.IllegalStateException: Illegal state. Please init first before obtaining provider
2024-11-08T10:39:56.1337184Z 	at org.keycloak.common.crypto.CryptoIntegration.getProvider(CryptoIntegration.java:48)
2024-11-08T10:39:56.1337949Z 	at org.keycloak.common.util.PemUtils.generateThumbprint(PemUtils.java:135)
2024-11-08T10:39:56.1338619Z 	at org.keycloak.jose.jwk.JWK.setX509CertificateChain(JWK.java:125)
2024-11-08T10:39:56.1339430Z 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
2024-11-08T10:39:56.1340206Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2024-11-08T10:39:56.1340987Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:141)
2024-11-08T10:39:56.1341643Z 	... 19 more
```

The reason is that the json properties `x5t` and `x5t#S256` are now calculated directly when reading the keys and that needs the `CryptoIntegration`. I think we have to only calculate those properties if they are requested.

### Version

999

### Regression

- [X] The issue is a regression

### Expected behavior

The keycloak-client needs to read the `JSONWebKeySet` without needing the `CryptoIntegration`.

### Actual behavior

Exception thrown.

### How to Reproduce?

Sync the keycloak client with the script `sync-keycloak-sources.sh` with  the current code and execute the test.

```
mvn -f testsuite/admin-client-tests/pom.xml test -Dkeycloak.version.docker.image=nightly -Dtest=GroupMappersTest#testGroupMappers
```

### Anything else?

_No response_

--- a/core/src/main/java/org/keycloak/jose/jwk/JWK.java
+++ b/core/src/main/java/org/keycloak/jose/jwk/JWK.java
@@ -75,8 +75,10 @@ public class JWK {
     @JsonProperty(X5C)
     private String[] x509CertificateChain;
 
+    @JsonProperty(SHA1_509_THUMBPRINT)
     private String sha1x509Thumbprint;
 
+    @JsonProperty(SHA256_509_THUMBPRINT)
     private String sha256x509Thumbprint;
 
     protected Map<String, Object> otherClaims = new HashMap<String, Object>();
@@ -120,24 +122,36 @@ public class JWK {
 
     public void setX509CertificateChain(String[] x509CertificateChain) {
         this.x509CertificateChain = x509CertificateChain;
-        if (x509CertificateChain != null && x509CertificateChain.length > 0) {
-            try {
-              sha1x509Thumbprint = PemUtils.generateThumbprint(x509CertificateChain, "SHA-1");
-              sha256x509Thumbprint = PemUtils.generateThumbprint(x509CertificateChain, "SHA-256");
-            } catch (NoSuchAlgorithmException e) {
-              throw new RuntimeException(e);
-            }
-        }
     }
 
-    @JsonProperty(SHA1_509_THUMBPRINT)
     public String getSha1x509Thumbprint() {
+        if (sha1x509Thumbprint == null && x509CertificateChain != null && x509CertificateChain.length > 0) {
+            try {
+                sha1x509Thumbprint = PemUtils.generateThumbprint(x509CertificateChain, "SHA-1");
+            } catch (NoSuchAlgorithmException e) {
+                throw new RuntimeException(e);
+            }
+        }
         return sha1x509Thumbprint;
     }
 
-    @JsonProperty(SHA256_509_THUMBPRINT)
+    public void setSha1x509Thumbprint(String sha1x509Thumbprint) {
+        this.sha1x509Thumbprint = sha1x509Thumbprint;
+    }
+
     public String getSha256x509Thumbprint() {
+        if (sha256x509Thumbprint == null && x509CertificateChain != null && x509CertificateChain.length > 0) {
+            try {
+                sha256x509Thumbprint = PemUtils.generateThumbprint(x509CertificateChain, "SHA-256");
+            } catch (NoSuchAlgorithmException e) {
+                throw new RuntimeException(e);
+            }
+        }
         return sha256x509Thumbprint;
+    }
+
+    public void setSha256x509Thumbprint(String sha256x509Thumbprint) {
+        this.sha256x509Thumbprint = sha256x509Thumbprint;
     }
 
     @JsonAnyGetter

